library/ssl_tls13_keys.c

```diff
@@ -1052,6 +1058,194 @@ int mbedtls_ssl_tls13_populate_transform( mbedtls_ssl_transform *transform,
+ * the early application data and handshake messages as described in section 7
+ * of RFC 8446.
+ *
+ * NOTE: Only one key is generated, the key for the traffic from the client to
+ *       the server. The TLS 1.3 specification does not define a secret and thus
+ *       a key for server early traffic.
+ */
+MBEDTLS_CHECK_RETURN_CRITICAL
+static int ssl_tls13_generate_early_key( mbedtls_ssl_context *ssl,
+                                         mbedtls_ssl_key_set *traffic_keys )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+
+    mbedtls_md_type_t md_type;
+
+    psa_algorithm_t hash_alg;
+    size_t hash_len;
+
+    unsigned char transcript[MBEDTLS_TLS1_3_MD_MAX_SIZE];
+    size_t transcript_len;
+
+    size_t key_len, iv_len;
+
+    mbedtls_ssl_handshake_params *handshake = ssl->handshake;
+    const mbedtls_ssl_ciphersuite_t *ciphersuite_info = handshake->ciphersuite_info;
+    mbedtls_ssl_tls13_early_secrets *tls13_early_secrets = &handshake->tls13_early_secrets;
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> ssl_tls13_generate_early_key" ) );
+
+    ret = ssl_tls13_get_cipher_key_info( ciphersuite_info, &key_len, &iv_len );
+    if( ret != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 1, "ssl_tls13_get_cipher_key_info", ret );
+        return ret;
+    }
+
+    md_type = ciphersuite_info->mac;
+
+    hash_alg = mbedtls_hash_info_psa_from_md( ciphersuite_info->mac );
+    hash_len = PSA_HASH_LENGTH( hash_alg );
+
+    ret = mbedtls_ssl_get_handshake_transcript( ssl, md_type,
+                                                transcript,
+                                                sizeof( transcript ),
+                                                &transcript_len );
+    if( ret != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 1,
+                               "mbedtls_ssl_get_handshake_transcript",
+                               ret );
+        return( ret );
+    }
+
+    ret = mbedtls_ssl_tls13_derive_early_secrets(
+              hash_alg, handshake->tls13_master_secrets.early,
+              transcript, transcript_len, tls13_early_secrets );
+    if( ret != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET(
+            1, "mbedtls_ssl_tls13_derive_early_secrets", ret );
+        return( ret );
+    }
+
+    MBEDTLS_SSL_DEBUG_BUF(
+        4, "Client early traffic secret",
+        tls13_early_secrets->client_early_traffic_secret, hash_len );
+
+    /*
+     * Export client handshake traffic secret
+     */
+    if( ssl->f_export_keys != NULL )
+    {
+        ssl->f_export_keys(
+            ssl->p_export_keys,
+            MBEDTLS_SSL_KEY_EXPORT_TLS1_3_CLIENT_EARLY_SECRET,
+            tls13_early_secrets->client_early_traffic_secret,
+            hash_len,
+            handshake->randbytes,
+            handshake->randbytes + MBEDTLS_CLIENT_HELLO_RANDOM_LEN,
+            MBEDTLS_SSL_TLS_PRF_NONE /* TODO: FIX! */ );
+    }
+
+    ret = ssl_tls13_make_traffic_key(
```

Comment:
After this call shouldn't we zeroize tls13_early_secrets or at least `tls13_early_secrets->client_early_traffic_secret` ? I does not seem to be done for `tls13_hs_secrets` either but I'd say that's a miss.

Reply:
`tls13_hs_secrets` is clean at ln 788. But I still think there are some issues.

Reply:
Done for both. 

I prepare to do some investigation on it to make sure below issues. For time being, I am not sure if they should be fixed in this PR.
- `mbedtls_ssl_tls13_early_secrets` is only used in `ssl_tls13_generate_early_key`, it can be removed from `ssl->handshake`
   - `mbedtls_ssl_tls13_derive_early_secrets` is only called in `ssl_tls13_generate_early_key`
- `tls13_hs_secrets` is zeroized in `mbedtls_ssl_tls13_calculate_verify_data` which is called in finished message.  I want to make sure where the output is used and if we can optimized it.

Reply:
>     * `mbedtls_ssl_tls13_early_secrets` is only used in `ssl_tls13_generate_early_key`, it can be removed from `ssl->handshake`
>       
>       * `mbedtls_ssl_tls13_derive_early_secrets` is only called in `ssl_tls13_generate_early_key`

It MUST change many lines. Improve it in future.
> 
>     * `tls13_hs_secrets` is zeroized in `mbedtls_ssl_tls13_calculate_verify_data` which is called in finished message.  I want to make sure where the output is used and if we can optimized it.

It must change the structure of `mbedtls_ssl_tls13_{write,process}_finished_message` . Improve it in future.

Reply:
I just revert the `erase tls13_hs_secrets` . And it is erased in `mbedtls_ssl_tls13_calculate_verify_data` .